relayer/chain/ethereum/message.go

```diff
@@ -6,156 +6,52 @@ package ethereum
 	"github.com/snowfork/polkadot-ethereum/relayer/chain"
+	"github.com/snowfork/polkadot-ethereum/relayer/contracts/outbound"
+	"github.com/snowfork/polkadot-ethereum/relayer/substrate"
 )
 
-type Message struct {
-	Data              []byte
-	VerificationInput VerificationInput
-}
-
-type VerificationInput struct {
-	IsBasic        bool
-	AsBasic        VerificationBasic
-	IsReceiptProof bool
-	AsReceiptProof VerificationReceiptProof
-	IsNone         bool
-}
-
-type VerificationBasic struct {
-	BlockNumber uint64
-	EventIndex  uint32
-}
-
-type VerificationReceiptProof struct {
-	BlockHash types.H256
-	TxIndex   types.U32
-	Proof     *Proof
-}
-
-type Proof struct {
-	Keys   []types.Bytes
-	Values []types.Bytes
-}
-
-func NewProof() *Proof {
-	return &Proof{
-		Keys:   make([]types.Bytes, 0),
-		Values: make([]types.Bytes, 0),
-	}
-}
-
-// For interface ethdb.KeyValueWriter
-func (p *Proof) Put(key []byte, value []byte) error {
-	p.Keys = append(p.Keys, types.NewBytes(common.CopyBytes(key)))
-	p.Values = append(p.Values, types.NewBytes(common.CopyBytes(value)))
-	return nil
-}
-
-// For interface ethdb.KeyValueWriter
-func (p *Proof) Delete(_ []byte) error {
-	return fmt.Errorf("Delete should never be called to generate a proof")
-}
-
-func (v VerificationInput) Encode(encoder scale.Encoder) error {
-	var err1, err2 error
-	switch {
-	case v.IsBasic:
-		err1 = encoder.PushByte(0)
-		err2 = encoder.Encode(v.AsBasic)
-	case v.IsReceiptProof:
-		err1 = encoder.PushByte(1)
-		err2 = encoder.Encode(v.AsReceiptProof)
-	case v.IsNone:
-		err1 = encoder.PushByte(2)
-	default:
-		return fmt.Errorf("VerificationInput must be one of the corresponding Rust enum types")
-	}
-
-	if err1 != nil {
-		return err1
-	}
-	if err2 != nil {
-		return err2
-	}
-
-	return nil
-}
-
-func (v *VerificationInput) Decode(decoder scale.Decoder) error {
-	tag, err := decoder.ReadOneByte()
-	if err != nil {
-		return err
-	}
-
-	switch tag {
-	case 0:
-		v.IsBasic = true
-		err = decoder.Decode(&v.AsBasic)
-	case 1:
-		v.IsReceiptProof = true
-		err = decoder.Decode(&v.AsReceiptProof)
-	case 2:
-		v.IsNone = true
-	default:
-		return fmt.Errorf("VerificationInput must be one of the corresponding Rust enum types")
-	}
-
-	if err != nil {
-		return err
-	}
-
-	return nil
-}
-
-func MakeMessageFromEvent(event *etypes.Log, receiptsTrie *etrie.Trie, log *logrus.Entry) (*chain.Message, error) {
+func MakeMessageFromEvent(event *outbound.ContractMessage, receiptsTrie *etrie.Trie, log *logrus.Entry) (*chain.EthereumOutboundMessage, error) {
 	// RLP encode event log's Address, Topics, and Data
 	var buf bytes.Buffer
-	err := event.EncodeRLP(&buf)
+	err := event.Raw.EncodeRLP(&buf)
 	if err != nil {
 		return nil, err
 	}
 
-	receiptKey, err := rlp.EncodeToBytes(event.TxIndex)
+	receiptKey, err := rlp.EncodeToBytes(event.Raw.TxIndex)
 	if err != nil {
 		return nil, err
 	}
 
-	proof := NewProof()
+	proof := substrate.NewMerkleProof()
```

Comment:
I'd personally drop the "merkle" because we have the header proofs around as well - also technically merkle proofs - but formatted and validated using very different code. The "merkle" commonality doesn't translate to any shared code in substrate or the relayer

Reply:
Ditto on the substrate side

Reply:
ah good to know. Since its generic proof data, how about this renaming?:

```diff
pub struct Message {
	pub data: Vec<u8>,
	pub proof: Proof,
}

pub struct Proof {
	pub block_hash: H256,
	pub tx_index: u32,
-	pub merkle_proof: (Vec<Vec<u8>>, Vec<Vec<u8>>),
+	pub data: (Vec<Vec<u8>>, Vec<Vec<u8>>),
}

Reply:
Fixed. Renamed as described above.